### PR TITLE
Support kubelet to provide client cert for liveness/readiness probe check

### LINKED ISSUE #60952: Provide client cert for liveness/readiness probe check
Currently liveness/readiness probe only support http/TLS, but not mTLS, which blocks us from using this feature for mTLS services. 

We should provide option to let kubelet send client cert to services.

### PR DESCRIPTION
**Pull request by JimmyCYJ**
**What this PR does / why we need it**: Support kubelet to send HTTPS probe check and provide client cert in the check.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #60952 

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/probe/http/kubelet-ca.crt [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+EjEQMA4GA1UEChMHQWNtZSBDbzCBmzAQBgcqhkjOPQIBBgUrgQQAIwOBhgAEAXiJ
+Z8ceJNJsRnSdhl/ocPrOfCjQ3OcHDjraQilbnozn+h4fZoiwfLEBdouphRH0ZBx2
+akRUNvPqf6TK1ZUGEdXbABxOFrx68rASn0bYF46vTXpfl6P85ocP/gQVPoq3/IVK
+3/aR7t5PtJY4ckXOYcQ+ihpTE3KMp+V2rMHa0sGrxRESo2EwXzAOBgNVHQ8BAf8E
+BAMCAqQwHQYDVR0lBBYwFAYIKwYBBQUHAwEGCCsGAQUFBwMCMA8GA1UdEwEB/wQF
+MAMBAf8wHQYDVR0RBBYwFIIJbG9jYWxob3N0gQdhQGEuY29tMAoGCCqGSM49BAME
+A4GMADCBiAJCAWd0oBAIVSX4N3gyUUH17hu6zYqe3DgRdIXY0ciIR7fZu7pf7zQM
+W8tWPoL+aMGahhWnJexG4+SxWeFLA8FdFA4uAkIBa3UCCSC6AUI/ZNXt3A87C+GY
++cvRAQlAb0ryRBu1HaXLvMn0QiCfGWwi4PBmApdkvqyjfLrec3GqC96LA+9Kumw=
+-----END CERTIFICATE-----
```
Comment (wattli) [inline pkg/probe/http/kubelet-ca.crt]: Add a new line to the end. Same below

File: pkg/kubelet/prober/common_test.go [hunk 2]
```diff
@@ -108,6 +108,7 @@ func newTestManager() *manager {
 		nil, // runner
 		refManager,
 		&record.FakeRecorder{},
+		"",
```
Comment (wattli) [inline pkg/kubelet/prober/common_test.go]: add some comment about the meaning of this empty string?

**Review by wattli — APPROVED** [review]
/lgtm, thanks!

File: pkg/probe/http/kubelet-ca.crt (hunk 1 above)
Comment (JimmyCYJ) [reply pkg/probe/http/kubelet-ca.crt]: Done. Thanks.

File: pkg/kubelet/prober/common_test.go (hunk 2 above)
Comment (JimmyCYJ) [reply pkg/kubelet/prober/common_test.go]: Added comment. Thanks.

File: pkg/probe/http/http.go [hunk 3]
```diff
@@ -22,20 +22,67 @@ import (
...
 
+	"crypto/x509"
 	"github.com/golang/glog"
 )
 
-func New() HTTPProber {
+const (
+	defaultKubeletClientCertificateFile = "kubelet-client.crt"
+	defaultKubeletClientKeyFile         = "kubelet-client.key"
+	defaultKubeletCACertificateFile     = "kubelet-ca.crt"
```
Comment (liggitt) [inline pkg/probe/http/http.go]: hard-coded files like this aren't appropriate... I'd expect these to come from kubelet config

Comment (liggitt) [reply pkg/probe/http/http.go]: also, it's unclear how this connects to the client credential management the kubelet has (client cert rotation, initialized from the `--kubeconfig` arg, etc)

**Comment by liggitt** [comment]
are we expecting the kubelet to send a client cert intended to authenticate it to the apiserver to all liveness checks as well?

File: pkg/probe/http/http.go [hunk 4]
```diff
@@ -22,20 +22,67 @@ import (
...
+	if err != nil {
+		glog.Errorf("unable to read CA cert: %v", err)
+		return NewHttpProber()
+	}
+	certPool := x509.NewCertPool()
+	certPool.AppendCertsFromPEM(caCert)
+
+	tlsConfig := &tls.Config{
+		Certificates:       []tls.Certificate{cert},
+		RootCAs:            certPool,
```
Comment (lookuptable) [inline pkg/probe/http/http.go]: I don't think you need to set `RootCAs` since `InsecureSkipVerify` is set to true

File: pkg/probe/http/http.go [hunk 5]
```diff
@@ -22,20 +22,67 @@ import (
...
+
+	certPath, err := filepath.Abs(filepath.Join(certDir, defaultKubeletClientCertificateFile))
+	if err != nil {
+		glog.Errorf("unable to build HTTPS prober cert path: %v", err)
+		return NewHttpProber()
+	}
+	keyPath, err := filepath.Abs(filepath.Join(certDir, defaultKubeletClientKeyFile))
+	if err != nil {
+		glog.Errorf("unable to build HTTPS prober key path: %v", err)
+		return NewHttpProber()
```
Comment (lookuptable) [inline pkg/probe/http/http.go]: Will it be confusing to return a HttpProber whenever an error occurs?

**Comment by wattli** [comment]
_are we expecting the kubelet to send a client cert intended to authenticate it to the apiserver to all liveness checks as well?_

@liggit, this is a good question, but it seems to be the only thing available for now, do you have any suggestions?

File: pkg/probe/http/http.go (hunk 4 above)
Comment (wattli) [reply pkg/probe/http/http.go]: That's right.

Comment (JimmyCYJ) [reply pkg/probe/http/http.go]: Thanks for pointing that out. Removed RootCA

File: pkg/probe/http/http.go (hunk 5 above)
Comment (JimmyCYJ) [reply pkg/probe/http/http.go]: Removed. In this case, we still create a HTTPSProber, and the probe will fail.

**Comment by wattli** [comment]
@liggit, since cert is some public info, it's up to the receiver to accept it or not.

**Comment by JimmyCYJ** [comment]
Please take a look. Thanks!

**Comment by krmayankk** [comment]
>  are we expecting the kubelet to send a client cert intended to authenticate it to the apiserver to all liveness checks as well?

@liggitt @wattli this is interesting from a multi tenancy point of view. We are saying the kubelet will use the same client cert to talk to both apiserver and the tenants pod container ?  I am not an expert, but does this decrease the apiserver security in any respect ?

**Review by tallclair — COMMENTED** [review]
There's a security issue in this PR, which unfortunately I think might be a dead end. HTTPGetAction probes can be pointed at an arbitrary host, so any user that can create a Pod can make an arbitrary get request authenticated as the Kubelet.

If checking an authenticated health endpoint is a strict requirement, I think the recommended workaround for now would be to run a sidecar that is capable of making the authenticated request, and issue that via an exec probe from the Kubelet.

**Comment by PiotrSikora** [comment]
@tallclair as far as I understand, your concern is only about re-using the client certificate that kubelet uses to authenticate within K8s, is that right? If so, would this change be acceptable if kubelet used a different (i.e. user provided) client certificate for the HTTPS health checks?

**Comment by blademainer** [comment]
mark

**Pull request closed without merging**